### ISSUE TITLE
NMS

### ISSUE THREAD

**Issue by AidanAbramson** [ultralytics/ultralytics#643]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Is there any layer of NMS in the model? YOLOv7 had something like that but when I predict with my custom trained model it sometimes produces 2-5 detections on the same object... all with similar confidence...

**Comment by AidanAbramson** [comment]
@medphisiker 
> I also convert to onnx format YOLOv8n models and visualise them by https://netron.app/. There is no NMS layers in the models ) 

Yeah that exactly what I've done. Does that mean I'll have to manually add an NMS plugin when exporting?

@stastatun Could you let me know where is that line of code written? Maybe I have an outdated version...

Thanks so much for the replies!

**Comment by medphisiker** [comment]
@AidanAbramson - Does that mean I'll have to manually add an NMS plugin when exporting?

Unfortunately, I don't have an exact answer to this question. Perhaps the developers will also add or have already added the ability to export the model to `onnx `with `NMS `as part of the model.

Initially, there is also no `NMS `as part of the `YOLOv7 ` model , it is attached to the model during the export process to `onnx ` model. Perhaps you can adapt the code from `YOLOv7 `to `YOLOv8`.
I haven't done it myself yet and I don't know how difficult this process is.

I will be glad to know how this process will go.
In one of the issue ([link](https://github.com/ultralytics/ultralytics/issues/480)), the developers planned to add the output of ready-made masks to the model. 
Perhaps in the future, the ability to export `YOLOv8 `with NMS as part of the model like for `YOLOv7 `will be added =)

**Comment by stastatun** [comment]
@AidanAbramson Also not a developer, just a user, but here's what I've gathered so far:
When a `DetectionModel` is [called](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/yolo/engine/model.py#L64), it runs the inference in [BasePredictor.stream_inference](https://github.com/ultralytics/ultralytics/blob/15b3b0365ab2f12993a58985f3cb7f2137409a0c/ultralytics/yolo/engine/predictor.py#L166) and then does post-processing in [DetectionPredictor.postprocess](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/yolo/v8/detect/predict.py#L23), which includes NMS.
It probably isn't a lot of work to add NMS in [Detect.forward](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/nn/modules.py#L400) so it will export with the model, although doing so without hurting the performance isn't trivial.

@medphisiker I don't think adding NMS to the exported ONNX model would be too difficult, but doing so without degrading the inference speed definitely is not trivial. Pretty sure the best way to go about would be to modify the graph to use [EfficientNMS_TRT](https://github.com/NVIDIA/TensorRT/tree/main/plugin/efficientNMSPlugin) instead of `torchvision.ops.nms`.

**Comment by AidanAbramson** [comment]
@stastatun Ok I get what you're saying. Thanks for the reply! I rather not change the original when it comes to the performance of the model... Hope this would be implemented in the future as it's rather an important layer...

About EfficientNMS_TRT, it seems not to work with a TRT version under 8.4 (or at least with YOLOv7 but I doubt it's an architecture thing). Would you recommend using BatchedNMS instead?

**Comment by stastatun** [comment]
@AidanAbramson Investigated a little further, and looks like TensorRT actually [does support regular NMS](https://github.com/onnx/onnx-tensorrt/blob/main/docs/operators.md)! So I it could be worth a shot to just add the postprocessing function call in the forward pass of the model see if it exports. However I'd guess that this  _most likely_ won't work with segmentation tasks - only detection.

Don't really know the what's the difference between EfficientNMS and BatchedNMS.

**Comment by AidanAbramson** [comment]
@stastatun Oh that's great! Thanks so much for updating me :)

**Comment by triple-Mu** [comment]
https://github.com/triple-Mu/YOLOv8-TensorRT，
Here is a TensorRT solution

**Comment by pderrenger** [comment]
@triple-Mu thanks for sharing the TensorRT solution! It's great to see community contributions providing alternative implementations. If you have any questions or need further assistance, feel free to ask.

**Comment by laugh12321** [comment]
Hello everyone!

I would like to introduce my open-source project - [TensoRT-YOLO](https://github.com/laugh12321/TensorRT-YOLO), a tool for deploying YOLO Series (Support YOLOv9) with Efficient NMS in TensorRT. 

## Key Features

- Supports FLOAT32, FLOAT16 ONNX export, and TensorRT inference
- Supports YOLOv5, YOLOv8, YOLOv9, PP-YOLOE, and PP-YOLOE+
- Integrates EfficientNMS TensorRT plugin for accelerated post-processing
- Utilizes CUDA kernel functions to accelerate preprocess
- Supports C++ and Python inference

## Perfomance

Perfomance Test using GPU RTX 2080Ti 22GB on AMD Ryzen 7 5700X 8-Core/ 128GB RAM.

Model Performance Evaluation using TensorRT engine using [TensoRT-YOLO](https://github.com/laugh12321/TensorRT-YOLO).

All models were deployed using FP16, BatchSize 4 and size 640. 

### YOLOv9 Series

This includes the YOLOv9-C, YOLOv9-E, YOLOv9-C-Converted, YOLOv9-E-Converted, GELAN-C and GELAN-E.

| YOLOv9-C           | YOLOv9-E           | YOLOv9-C-Converted | YOLOv9-E-Converted | GELAN-C | GELAN-E |
| ------------------ | ------------------ | ------------------ | ------------------ | ------- | ------- |
| Average Latency: `36.615ms` | Average Latency: `59.736ms` | Average Latency: `19.689ms` | Average Latency: `53.144ms` | Average Latency: `19.557ms` | Average Latency: `53.575ms` |

### YOLOv8 Series

This includes the YOLOv8n, YOLOv8s, YOLOv8m, YOLOv8l and YOLOv8x.

| YOLOv8n           | YOLOv8s          | YOLOv8m | YOLOv8l | YOLOv8x |
| ------------------ | ------------------ | ------------------ | ------------------ | ------- |
| Average Latency: `10.289ms` | Average Latency: `12.459ms` | Average Latency: `18.514ms` | Average Latency: `24.926ms` | Average Latency: `34.587ms` |

**Comment by glenn-jocher** [comment]
@laugh12321 hello there! 🎉

This is fantastic! Your [TensoRT-YOLO](https://github.com/laugh12321/TensorRT-YOLO) project looks like a great resource for the community, especially with its support for the latest YOLO versions and the integration of EfficientNMS for TensorRT. The performance metrics you've shared are impressive and will surely be beneficial for those looking to deploy YOLO models more efficiently. 

It's great to see such contributions that help bridge the gap between research and real-world applications. Keep up the awesome work! 

For anyone looking into deploying YOLO models with TensorRT, this seems like a valuable tool to check out. Thanks for sharing! 🚀

**Comment by laugh12321** [comment]
@glenn-jocher Thank you so much for the kind words! 🌟 Your support means a lot.

**Comment by glenn-jocher** [comment]
@laugh12321 you're welcome! 😊 If you have any more questions or need further assistance, feel free to reach out. Happy coding!

**Comment by UltralyticsAssistant** [comment]
The associated pull request [#18484](https://github.com/ultralytics/ultralytics/pull/18484) has been merged, and it includes a potential fix that may resolve this issue! 🎉  

### Relevant Key Changes from the PR:
This update introduces extended export functionalities and post-processing improvements for object detection models, addressing several limitations in the workflow. Notable changes include:
- **Non-Maximum Suppression (NMS) Integration**: Exported models (TorchScript, ONNX, OpenVINO, TensorRT, TensorFlow) now include embedded NMS, providing seamless post-processing capabilities.
- **Enhanced Post-Processing Logic**: Predictions for various tasks (detection, segmentation, pose, and OBB) have been refined, including additional attribute support, like `keypoints`, for improved handling of prediction results.
- **Flexibility in Export Configurations**: Users can now toggle NMS during the export process, enabling easier deployment customization.
- **Better Metadata Handling**: Updates for improved export compatibility across formats.
- 🆙 Version upgraded to `8.3.67` for tracking this new functionality.

These changes aim to streamline the export process while improving runtime efficiency and usability for deployment scenarios.  

### Contributors ❤️
This valuable update was made possible due to contributions from [Y-T-G](https://github.com/Y-T-G), [lifafang](https://github.com/lifafang), [boorad](https://github.com/boorad), [glenn-jocher](https://github.com/glenn-jocher), [Laughing-q](https://github.com/Laughing-q), and [inisis](https://github.com/inisis). A big thank you to everyone involved! 🙌  

### How to Test the Changes:
To verify if this resolves your issue:
- Install the latest changes directly from the `main` branch:
  ```bash
  pip install git+https://github.com/ultralytics/ultralytics.git@main
  ```
- Or you can wait for the next official release on PyPI if preferred.

Let us know if this resolves the issue or if you encounter any remaining problems. Your feedback is invaluable in ensuring the robustness of these updates.  

Thanks again for bringing up this issue! 🙏 If anything persists or additional adjustments are needed, feel free to share more details.
